### ISSUE TITLE
iOS app crashes on creating Texture

### ISSUE THREAD

**Issue by Raki** [google/filament#814]
I'm trying to render a texture quad by following the "hello-triangle" example on iOS. My new material to render the quad is proper as it is drawing a black quad if I didn't bind any texture to material instance. But when I call this function to create texture, it is crashing at `generateMipmaps(*engine)`

`void loadTexture(Engine* engine, const std::string& filePath, Texture** map, bool sRGB)`
`{`
`       int w, h, n;`
        `unsigned char* data = stbi_load(path.c_str(), &w, &h, &n, 3);`
`        if (data != nullptr) {`
 `           *map = Texture::Builder()`
  `          .width(uint32_t(w))`
   `        .height(uint32_t(h))`
    `       .levels(0xff)`
     `      .format(sRGB?driver::TextureFormat::SRGB8:driver::TextureFormat::RGB8)`
      `      .build(*engine);`
       `     Texture::PixelBufferDescriptor buffer(data, size_t(w * h * 3),`
                                                 `Texture::Format::RGB, Texture::Type::UBYTE,`
                                                  `(driver::BufferDescriptor::Callback) &stbi_image_free);`
        `    (*map)->setImage(*engine, 0, std::move(buffer));`
         `   (*map)->generateMipmaps(*engine);`
`}`

I pulled this function from Desktop samples of filament.
I tried tweaking png compression options on xCode but no luck. 

 - Device: iPhone 7
 - OS: iOS 11.4.1

**Comment by bejado** [comment]
Which backend are you using, OpenGL? I think there were some changes recently to mipmap generation, but it's possible this is specific to iOS. Is there any particular message logged before crashing?

**Comment by Raki** [comment]
if format is **RGB8** then log is

019-02-14 10:35:51.922809+0530 FilamentDemo[291:12092] [DYMTLInitPlatform] platform initialization successful
FEngine (64 bits) created at 0x104800000 (threading is enabled)
FEngine resolved backend: OpenGL
2019-02-14 10:35:52.216054+0530 FilamentDemo[291:12101] Metal GPU Frame Capture Enabled
2019-02-14 10:35:52.217089+0530 FilamentDemo[291:12101] Metal API Validation Enabled
N5utils18PostconditionPanicE
in virtual void filament::PlatformCocoaTouchGL::makeCurrent(filament::driver::Platform::SwapChain *, filament::driver::Platform::SwapChain *):129
reason: Incomplete framebuffer.
#0 0   libsystem_pthread.dylib             0x0000000184cd1110 <redacted> + 0
libc++abi.dylib: terminating with uncaught exception of type utils::PostconditionPanic

if format is **SRGB8** then log is

2019-02-14 10:38:53.261018+0530 FilamentDemo[297:13117] [DYMTLInitPlatform] platform initialization successful
FEngine (64 bits) created at 0x109800000 (threading is enabled)
FEngine resolved backend: OpenGL
2019-02-14 10:38:53.542747+0530 FilamentDemo[297:13125] Metal GPU Frame Capture Enabled
2019-02-14 10:38:53.543834+0530 FilamentDemo[297:13125] Metal API Validation Enabled
N5utils18PostconditionPanicE
in void filament::details::FTexture::generateMipmaps(filament::details::FEngine &) const:199
reason: Texture format must be color renderable
#0 0   FilamentDemo                        0x0000000104e8b08c -[FilamentView initializeFilamentQuad] + 1388
#1 0   FilamentDemo                        0x0000000104e8a5f8 -[FilamentView initWithCoder:] + 180
#2 0   UIKit                               0x000000018f32cfec <redacted> + 248
#3 0   UIKit                               0x000000018f48e738 <redacted> + 688
#4 0   UIKit                               0x000000018f48e470 <redacted> + 312
#5 0   UIKit                               0x000000018f32cc94 <redacted> + 188
#6 0   UIKit                               0x000000018f48e738 <redacted> + 688
#7 0   UIKit                               0x000000018f48e8b0 <redacted> + 1064
#8 0   UIKit                               0x000000018f48e470 <redacted> + 312
#9 0   UIKit                               0x000000018f32bfe4 <redacted> + 1164
#10 0   UIKit                               0x000000018f15daa8 <redacted> + 372
#11 0   UIKit                               0x000000018eda1074 <redacted> + 176
#12 0   UIKit                               0x000000018ecc5b14 <redacted> + 172
#13 0   UIKit                               0x000000018ecc5a50 <redacted> + 28
#14 0   UIKit                               0x000000018ecb6ce4 <redacted> + 136
#15 0   UIKit                               0x000000018ecb5b18 <redacted> + 272
#16 0   UIKit                               0x000000018ed4362c <redacted> + 48
#17 0   UIKit                               0x000000018ecb9274 <redacted> + 3660
libc++abi.dylib: terminating with uncaught exception of type utils::PostconditionPanic

**Comment by bejado** [comment]
I don't think iOS supports RGB colorbuffers, hence the ["Texture format must be color renderable."](https://github.com/google/filament/blob/18bff7ef5e2acb8d465eba68a9fd2229f0529be2/filament/src/Texture.cpp#L198-L200) assertion. Can you try with a RGBA texture and see if that succeeds?

**Comment by Raki** [comment]
tried with format `driver::TextureFormat::RGBA8` got the following error

2019-02-14 10:47:16.609869+0530 FilamentDemo[302:15292] [DYMTLInitPlatform] platform initialization successful
FEngine (64 bits) created at 0x107800000 (threading is enabled)
FEngine resolved backend: OpenGL
2019-02-14 10:47:16.876139+0530 FilamentDemo[302:15301] Metal GPU Frame Capture Enabled
2019-02-14 10:47:16.877123+0530 FilamentDemo[302:15301] Metal API Validation Enabled
N5utils18PostconditionPanicE
in virtual void filament::PlatformCocoaTouchGL::makeCurrent(filament::driver::Platform::SwapChain *, filament::driver::Platform::SwapChain *):129
reason: Incomplete framebuffer.
#0 0   libsystem_pthread.dylib             0x0000000184cd1110 <redacted> + 0
libc++abi.dylib: terminating with uncaught exception of type utils::PostconditionPanic

with format `driver::TextureFormat::SRGBA8` got the following error

2019-02-14 10:49:25.732349+0530 FilamentDemo[308:16078] [DYMTLInitPlatform] platform initialization successful
FEngine (64 bits) created at 0x1037d8000 (threading is enabled)
FEngine resolved backend: OpenGL
2019-02-14 10:49:26.002348+0530 FilamentDemo[308:16086] Metal GPU Frame Capture Enabled
2019-02-14 10:49:26.003308+0530 FilamentDemo[308:16086] Metal API Validation Enabled
N5utils18PostconditionPanicE
in void filament::details::FTexture::generateMipmaps(filament::details::FEngine &) const:199
reason: Texture format must be color renderable

**Comment by bejado** [comment]
Okay, potentially something else going on- will take a closer look.

**Comment by romainguy** [comment]
This error triggers when `isRenderTargetFormatSupported()` returns `false` in the driver. Maybe it's just that it wasn't implemented properly for iOS?

**Comment by bejado** [comment]
@Raki just merged a fix- can pull and try again? Note that sRGB textures (correctly) do not support mipmap generation on iOS.

**Comment by Raki** [comment]
@bejado  ios build was crashing with the following errors


/../filament/src/driver/metal/MetalDriver.mm:450:56: error: no member named 'clear' in 'filament::driver::RenderPassParams'
    const auto clearFlags = (TargetBufferFlags) params.clear;
                                                ~~~~~~ ^
../../filament/src/driver/metal/MetalDriver.mm:482:21: error: no member named 'left' in 'filament::driver::RenderPassParams'
    viewport(params.left, params.bottom, params.width, params.height);
             ~~~~~~ ^
../../filament/src/driver/metal/MetalDriver.mm:482:34: error: no member named 'bottom' in 'filament::driver::RenderPassParams'
    viewport(params.left, params.bottom, params.width, params.height);
                          ~~~~~~ ^
../../filament/src/driver/metal/MetalDriver.mm:482:49: error: no member named 'width' in 'filament::driver::RenderPassParams'
    viewport(params.left, params.bottom, params.width, params.height);
                                         ~~~~~~ ^
../../filament/src/driver/metal/MetalDriver.mm:482:63: error: no member named 'height' in 'filament::driver::RenderPassParams'
    viewport(params.left, params.bottom, params.width, params.height);
                                                       ~~~~~~ ^
../../filament/src/driver/metal/MetalDriver.mm:672:88: error: no member named 'clear' in 'filament::driver::RenderPassParams'
    TargetBufferFlags clearFlags = (TargetBufferFlags) pImpl->mCurrentRenderPassParams.clear;
                                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
1 warning and 6 errors generated.
[219/228] Building CXX object filament/CMakeFiles/filament.dir/src/driver/vulkan/VulkanDriverImpl.cpp.o
ninja: build stopped: subcommand failed.

**Comment by Raki** [comment]
Hi @bejado , I tried with latest code, no crashes so far.
But the texture quad I drawing is coming like this.
![img_0236](https://user-images.githubusercontent.com/824691/52842687-e2cac000-3125-11e9-955d-084e35a9a8ce.PNG)

While debugging i can see like, the texture is being updated with current frame.
I'm using format as `driver::TextureFormat::RGBA8`

**Comment by bejado** [comment]
Does that only happen when you generate mipmaps? Can you show your material file?

**Comment by Raki** [comment]
@bejado , I tried disabling mipmaps but no change in output.
I'm using the following material .

`material`
 `                    `  `{`
    `                    `    ` name : simple_tex,`
    `                    ` `shadingModel : unlit,`
   `                    ` ` parameters : [`
     `                    ` `   // Base color sRGB texture`
      `                    `  ` {`
        `                    `     `type : sampler2d,`
       `                    `      `name : albedo`
      `                    ` `  }`
  `                    `  ` ],`
   `                    ` ` // To sample textures our material must declare that it requires`
   `                    `  `// a set of UV coordinates from the rendered mesh`
  `                    `  ` requires: [`
  `                    `       `uv0`
  `                    `   `]`
`}`

`fragment {`
    `void material(inout MaterialInputs material) {`
     `   prepareMaterial(material);`
 `       material.baseColor = texture(materialParams_albedo, getUV0());`
`    }`
`}`

**Comment by bejado** [comment]
I don't see anything wrong with the material. Are you creating a sampler and setting it as a parameter on your material instance?

```
TextureSampler sampler(TextureSampler::MinFilter::LINEAR, TextureSampler::MagFilter::LINEAR);
materialInstance = material->createInstance();
materialInstance->setParameter("albedo", texture, sampler);
```

It would help if you could post all the changes you've made to HelloTriangle.

**Comment by Raki** [comment]
@bejado The following are the changes I made to HelloTriangle Demo.


```
struct Vertex3
{
    math::float2 pos;
    math::float2 uv;
};

static const Vertex3 QUAD_VERTICES[4] = {
    {{-0.5,-0.5}, {0,0}},
    {{0.5,-0.5}, {1,0}},
{{0.5,0.5}, {1,1}},
{{-0.5,0.5}, {0,1}}
};

static constexpr uint16_t QUAD_INDICES[6] = {0,1,2,0,2,3};
```
```
-(void) initializeFilamentQuad
{
    engine = Engine::create(filament::Engine::Backend::OPENGL);
    swapChain = engine->createSwapChain((__bridge void*) self.layer);
    renderer = engine->createRenderer();
    scene = engine->createScene();
    camera = engine->createCamera();
    
    filaView = engine->createView();
    filaView->setClearColor({0.1, 0.125, 0.25, 1.0});
    filaView->setPostProcessingEnabled(false);
    filaView->setDepthPrepass(filament::View::DepthPrepass::DISABLED);
    
    int stride = sizeof(float)*4;
    
    app.vb = VertexBuffer::Builder()
    .vertexCount(4)
    .bufferCount(1)
    .attribute(VertexAttribute::POSITION, 0, VertexBuffer::AttributeType::FLOAT2, 0, stride)
    .attribute(VertexAttribute::UV0, 0, VertexBuffer::AttributeType::FLOAT2, 8, stride)
    .build(*engine);
    app.vb->setBufferAt(*engine, 0,
                        VertexBuffer::BufferDescriptor(QUAD_VERTICES, stride*4, nullptr));
    
    app.ib = IndexBuffer::Builder()
    .indexCount(6)
    .bufferType(IndexBuffer::IndexType::USHORT)
    .build(*engine);
    app.ib->setBuffer(*engine,
                      IndexBuffer::BufferDescriptor(QUAD_INDICES, sizeof(uint16_t)*6, nullptr));
    
    app.mat = Material::Builder()
    .package((void*) SIMPLE_TEXR_PACKAGE, sizeof(SIMPLE_TEXR_PACKAGE))
    .build(*engine);
    
    materialInstance = app.mat->getDefaultInstance();
    
    NSString *bcPath = [[NSBundle mainBundle] pathForResource:@"Model_1_txt" ofType:@"png"];
    const char *bcPath_cstr = [bcPath fileSystemRepresentation];

    TextureLoader::loadTexture(engine,string(bcPath_cstr),&albedo);
    TextureSampler *sampler = new TextureSampler(TextureSampler::MinFilter::LINEAR,TextureSampler::MagFilter::LINEAR);
    materialInstance->setParameter("albedo", albedo, *sampler);
    
    
    
    app.renderable = EntityManager::get().create();
    RenderableManager::Builder(1)
    .boundingBox({{ -1, -1, -1 }, { 1, 1, 1 }})
    .material(0, materialInstance)
    .geometry(0, RenderableManager::PrimitiveType::TRIANGLES, app.vb, app.ib, 0, 6)
    .culling(false)
    .receiveShadows(false)
    .castShadows(false)
    .build(*engine, app.renderable);
    scene->addEntity(app.renderable);
    
     
    filaView->setScene(scene);
    filaView->setCamera(camera);
    CGRect nativeBounds = [UIScreen mainScreen].nativeBounds;
    filaView->setViewport(Viewport(0, 0, nativeBounds.size.width, nativeBounds.size.height));
    
    
    constexpr float ZOOM = 1.5f;
    const uint32_t w = filaView->getViewport().width;
    const uint32_t h = filaView->getViewport().height;
    const float aspect = (float) w / h;
    camera->setProjection(Camera::Projection::ORTHO,
                          -aspect * ZOOM, aspect * ZOOM,
                          -ZOOM, ZOOM, 0, 1);
}
```

```
//loadTexture code 
        int w, h, n;
        unsigned char* data = stbi_load(path.c_str(), &w, &h, &n, 4);
        if (data != nullptr) {
            *map = Texture::Builder()
            .width(uint32_t(w))
            .height(uint32_t(h))
            .levels(0xff)
            .format(driver::TextureFormat::RGBA8)
            .build(*engine);
            Texture::PixelBufferDescriptor buffer(data, size_t(w * h * 4),
                                                  Texture::Format::RGBA, Texture::Type::UBYTE,
                                                  (driver::BufferDescriptor::Callback) &stbi_image_free);
            (*map)->setImage(*engine, 0, std::move(buffer));
            (*map)->generateMipmaps(*engine);
```

**Comment by bejado** [comment]
Can you try replacing

```
    materialInstance = app.mat->getDefaultInstance();
```

with

```
    materialInstance = app.mat->createInstance();
```

This might be a bug on our side.

**Comment by Raki** [comment]
@bejado Yes, it worked with   `app.mat->createInstance();` .  Thank you very much.

**Comment by bejado** [comment]
Sure thing! Note that you _should_ be able to use `getDefaultInstance` so that is in fact a bug on our part which will be fixed in #854.
